nodejs/aws-infra/experimental/loadBalancer.ts

```diff
@@ -0,0 +1,234 @@
+            // is needed here, but we anticipate this will not always be the case, so we include a
+            // set of values which must be unique.
+            const longName = `${name}-${containerName}-${portInfo.port}`;
+            const shortName = utils.sha1hash(`${longName}`);
+
+            // Create an internal load balancer if requested.
+            const cluster = portInfo.cluster;
+            const internal = cluster.network.usePrivateSubnets && !portInfo.external;
+
+            // See what kind of load balancer to create (application L7 for HTTP(S) traffic, or network L4 otherwise).
+            // Also ensure that we have an SSL certificate for termination at the LB, if that was requested.
+            const { listenerProtocol, targetProtocol, useAppLoadBalancer, certificateArn } =
+                computeLoadBalancerInfo(portInfo);
+
+            const parentOpts = { parent };
+
+            loadBalancer = new aws.elasticloadbalancingv2.LoadBalancer(shortName, {
+                ...loadBalancerArgs,
+                loadBalancerType: useAppLoadBalancer ? "application" : "network",
+                subnets: cluster.network.publicSubnetIds,
+                internal: internal,
+                // If this is an application LB, we need to associate it with the ECS cluster's security
+                // group, so that traffic on any ports can reach it.  Otherwise, leave blank, and
+                // default to the VPC's group.
+                securityGroups: useAppLoadBalancer ? [ cluster.instanceSecurityGroup.id ] : undefined,
+                tags: { Name: longName },
+            }, parentOpts);
+
+                    // Create the target group for the new container/port pair.
+            targetGroup = new aws.elasticloadbalancingv2.TargetGroup(shortName, {
+                port: portInfo.targetPort || portInfo.port,
+                protocol: targetProtocol,
+                vpcId: cluster.network.vpcId,
+                deregistrationDelay: 180, // 3 minutes
+                tags: { Name: longName },
+                targetType: "ip",
+            }, parentOpts);
+
+            // Listen on the requested port on the LB and forward to the target.
+            listener = new aws.elasticloadbalancingv2.Listener(longName, {
+                loadBalancerArn: loadBalancer.arn,
+                protocol: listenerProtocol,
+                certificateArn: certificateArn,
+                port: portInfo.port,
+                defaultAction: {
+                    type: "forward",
+                    targetGroupArn: targetGroup.arn,
+                },
+                // If SSL is used, we automatically insert the recommended ELB security policy from
+                // http://docs.aws.amazon.com/elasticloadbalancing/latest/application/create-https-listener.html.
+                sslPolicy: certificateArn ? "ELBSecurityPolicy-2016-08" : undefined,
+            }, parentOpts);
```

Comment:
@lukehoban Open questions from talking to @kennyjwilli. Basically, this approach can work in the sense that he can use this helpful default system, but fallback to just fully providing a full impl ILoadBalancerProvider himself.  However, tha'ts somewhat unpalatable as that's another large cliff to fall off of (even though it is at least good to have that escape hatch).

So, for our default interface implementations, we should still consider ways we can provide flexibility extensibility.  Perhaps by having these guys have their own optional args to help control things.